### PR TITLE
[TMP] Combine the sample connectionTestProgram and add-cluster-support-deux for cluster testing

### PR DESCRIPTION
**Pull request by mtwoodbury**
To allow the sample program to be run without getting the base branch add-cluster-support-deux, and then adding the connectionTestProgram branch. Created cluster-test and making it available. Can be temporary if the other two branches are used or can be a shortcut since it merges those two branches.

### CODE REVIEW

File: src/ReactiveDomain.Core/Util/Ensure.cs [hunk 1]
```diff
@@ -342,6 +342,7 @@ public static void GreaterThanOrEqualTo(decimal expected, decimal argument, stri
         public static void Between(int low, int high, int argument, string argumentName)
         {
             if (argument <= low || argument >= high)
+
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Core/Util/Ensure.cs]: We should wrap this with "{ }" for maintainability

File: src/ReactiveDomain.Core/Util/FluentExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace ReactiveDomain.Util
+{
+    public static class FluentExtensions {
+
+        public static T If<T>(this T t, Func<bool> cond, Func<T, T> builder) where T : class {
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Core/Util/FluentExtensions.cs]: Is there any more descriptive term we can use other than "T"? I would even be okay with "TObject"

File: src/ReactiveDomain.Persistence.Samples/IController.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace ReactiveDomain.Persistence.Samples {
+    public interface IController {
+
+        void Start();
+        void RunProcessingLoop();
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Persistence.Samples/IController.cs]: What is the pattern we're going for here? In other words, what is the difference between "Starting" and "Running the Processing Loop?" I would have expected something along the lines of "Start(); ... time marches on ... Stop();" with an IDisposable that made sure we were "Stopped".

File: src/ReactiveDomain.Persistence.Samples/TestCommandController.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                command = Console.ReadLine()?.Split(' '); // allow port and IP input in the future.
+
+                ReactiveDomain.UserCredentials userCredentials;
+                StreamStoreConnectionSettingsBuilder sscSettings;
+                EventStoreConnectionManager eventStoreConnectionManager;
+                IStreamStoreConnection streamStoreConnection;
+                // direct ES setting used by discover2 and gossip2
+                ConnectionSettings settings;
+
+                switch (command[0].ToLower()) {
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Persistence.Samples/TestCommandController.cs]: Wow, it would be a LOT easier to read this if we broke this up into methods. Very hard to follow.

File: src/ReactiveDomain.Persistence/EventStore/EventStoreConnectionManager.cs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        ///     <item>  
+        ///         <description>A cluster using a DNS name and HTTP Port</description>  
+        ///     </item>  
+        ///     <item>  
+        ///         <description>A cluster using one or more gossip seed IPAddresses and an HTTP Port</description>  
+        ///     </item>  
+        /// </list>  
+        /// </remarks>
+        /// <param name="connectionSettings"><see cref="StreamStoreConnectionSettings"/> defined by the caller.</param>
+        public EventStoreConnectionManager(StreamStoreConnectionSettings connectionSettings)
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Persistence/EventStore/EventStoreConnectionManager.cs]: This "feels" like something we could decompose more

File: src/ReactiveDomain.Persistence/EventStore/EventStoreLocalStartupUtils.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                        DirectoryInfo installPath,
+                        string args,
+                        UserCredentials credentials,
+                        IPAddress server,
+                        int tcpPort,
+                        ProcessWindowStyle windowStyle,
+                        StartConflictOption opt) {
+            Ensure.NotNullOrEmpty(args, "args");
+            Ensure.NotNull(credentials, "credentials");
+
```
Comment (skentjoneslinedata) [inline src/ReactiveDomain.Persistence/EventStore/EventStoreLocalStartupUtils.cs]: There is a lot going on in here. We may want to break this down further

**Comment by mtwoodbury** [comment]
Adding Tim and Daniel. This pull request is the same as https://github.com/linedata/reactive-domain/pull/65
This consolidates all the checkins from that PR into one for easier viewing. It also would make the merge easier I think. Any questions ping me.
I will add you to the original as well.
